tools/yulPhaser/Mutations.cpp

```diff
@@ -0,0 +1,147 @@
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <tools/yulPhaser/Mutations.h>
+
+#include <tools/yulPhaser/SimulationRNG.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <string>
+#include <vector>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::phaser;
+
+function<Mutation> phaser::geneRandomisation(double _chance)
+{
+	return [=](Chromosome const& _chromosome)
+	{
+		vector<string> optimisationSteps;
+		for (auto const& step: _chromosome.optimisationSteps())
+			optimisationSteps.push_back(
+				SimulationRNG::bernoulliTrial(_chance) ?
+				Chromosome::randomOptimisationStep() :
+				step
+			);
+
+		return Chromosome(move(optimisationSteps));
+	};
+}
+
+function<Mutation> phaser::geneDeletion(double _chance)
+{
+	return [=](Chromosome const& _chromosome)
+	{
+		vector<string> optimisationSteps;
+		for (auto const& step: _chromosome.optimisationSteps())
+			if (!SimulationRNG::bernoulliTrial(_chance))
+				optimisationSteps.push_back(step);
+
+		return Chromosome(move(optimisationSteps));
+	};
+}
+
+function<Mutation> phaser::geneAddition(double _chance)
+{
+	return [=](Chromosome const& _chromosome)
+	{
+		vector<string> optimisationSteps;
+
+		if (SimulationRNG::bernoulliTrial(_chance))
+			optimisationSteps.push_back(Chromosome::randomOptimisationStep());
+
+		for (auto const& step: _chromosome.optimisationSteps())
+		{
+			optimisationSteps.push_back(step);
+			if (SimulationRNG::bernoulliTrial(_chance))
+				optimisationSteps.push_back(Chromosome::randomOptimisationStep());
+		}
+
+		return Chromosome(move(optimisationSteps));
+	};
+}
+
+function<Mutation> phaser::alternativeMutations(
+	double _firstMutationChance,
+	function<Mutation> _mutation1,
+	function<Mutation> _mutation2
+)
+{
+	return [=](Chromosome const& _chromosome)
+	{
+		if (SimulationRNG::bernoulliTrial(_firstMutationChance))
+			return _mutation1(_chromosome);
+		else
+			return _mutation2(_chromosome);
+	};
+}
+
+namespace
+{
+
+Chromosome buildChromosomesBySwappingParts(
+	Chromosome const& _chromosome1,
+	Chromosome const& _chromosome2,
+	size_t _crossoverPoint
+)
+{
+	assert(_crossoverPoint <= _chromosome1.length());
+	assert(_crossoverPoint <= _chromosome2.length());
+
+	auto begin1 = _chromosome1.optimisationSteps().begin();
+	auto begin2 = _chromosome2.optimisationSteps().begin();
+
+	return Chromosome(
+		vector<string>(begin1, begin1 + _crossoverPoint) +
+		vector<string>(begin2 + _crossoverPoint, _chromosome2.optimisationSteps().end())
+	);
+}
+
+}
+
+function<Crossover> phaser::randomPointCrossover()
+{
+	return [=](Chromosome const& _chromosome1, Chromosome const& _chromosome2)
+	{
+		size_t minLength = min(_chromosome1.length(), _chromosome2.length());
+
+		// Don't use position 0 (because this just swaps the values) unless it's the only choice.
```

Comment:
Again, I don't think it is too bad - after all, it is random.

Reply:
Yeah, that's probably not all that important but it shouldn't hurt either and this is how it's done in the typical one-point crossover (e.g. as shown in the book "Genetic Algorithms + Data Structures = Evolution Programs" by Michalewicz, page 35).

The added complication for me here is that the classic algorithm works on fixed-length chromosomes and our chromosomes have variable length so I have no choice but to give up if one of the chromosomes is empty (which shouldn't be common but could happen if most sequences make the programs worse; I've seen it happen when running on multiple very short programs from the repo).